### ISSUE TITLE
Doors no longer fail in dCache 5.2

### ISSUE THREAD

**Issue by alrossi** [dCache/dcache#5308]
Refer to RT #9853: Doors no longer fail in dCache 5.2

The problem presents itself when there is an error during Spring bean creation.  This is not specific to any cell.   

Instead of the reported xrootd door problem, caused by a VOMSError, I have simulated the issue by adding a faulty bean to the alarms.xml config:

```
<!-- ALR FIXME -->
    <bean id="fake" class="org.dcache.alarms.file.FileBackedAlarmPriorityMap">
        <property name="propertiesPath" value="${some.phony.property}"/>
        <property name="defaultPriority" value="${some.other.phony.property}"/>
    </bean>
```

Upon domain start, this provokes the same stack trace:
```
07 Feb 2020 13:53:06 (Xrootd0-fndcatemp1) [] Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanDefinitionStoreException: Invalid b
ean definition with name 'fake' defined in class path resource [org/dcache/xrootd/door/xrootd.xml]: Could not resolve placeholder 'some-phony-property' in value "${some-phony-property}"; nested except
ion is java.lang.IllegalArgumentException: Could not resolve placeholder 'some-phony-property' in value "${some-phony-property}"
07 Feb 2020 13:53:06 (System) [] URL [file:/usr/share/dcache/services/xrootd.batch]: line 51: Command failed: (3) Invalid bean definition with name 'fake' defined in class path resource [org/dcache/xr
ootd/door/xrootd.xml]: Could not resolve placeholder 'some-phony-property' in value "${some-phony-property}"; nested exception is java.lang.IllegalArgumentException: Could not resolve placeholder 'som
e-phony-property' in value "${some-phony-property}"; caused by: org.springframework.beans.factory.BeanDefinitionStoreException: Invalid bean definition with name 'fake' defined in class path resource 
[org/dcache/xrootd/door/xrootd.xml]: Could not resolve placeholder 'some-phony-property' in value "${some-phony-property}"; nested exception is java.lang.IllegalArgumentException: Could not resolve pl
aceholder 'some-phony-property' in value "${some-phony-property}"
07 Feb 2020 13:53:06 (System) []        at org.springframework.beans.factory.config.PlaceholderConfigurerSupport.doProcessProperties(PlaceholderConfigurerSupport.java:228)
07 Feb 2020 13:53:06 (System) []        at org.springframework.context.support.PropertySourcesPlaceholderConfigurer.processProperties(PropertySourcesPlaceholderConfigurer.java:182)
07 Feb 2020 13:53:06 (System) []        at org.springframework.context.support.PropertySourcesPlaceholderConfigurer.postProcessBeanFactory(PropertySourcesPlaceholderConfigurer.java:157)
07 Feb 2020 13:53:06 (System) []        at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:286)
07 Feb 2020 13:53:06 (System) []        at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:166)
07 Feb 2020 13:53:06 (System) []        at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:706)
07 Feb 2020 13:53:06 (System) []        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:532)
07 Feb 2020 13:53:06 (System) []        at org.dcache.cells.UniversalSpringCell.createContext(UniversalSpringCell.java:1075)
07 Feb 2020 13:53:06 (System) []        at org.dcache.cells.UniversalSpringCell.starting(UniversalSpringCell.java:242)
07 Feb 2020 13:53:06 (System) []        at dmg.cells.nucleus.CellAdapter.prepareStartup(CellAdapter.java:722)
07 Feb 2020 13:53:06 (System) []        at dmg.cells.nucleus.CellNucleus.doStart(CellNucleus.java:931)
07 Feb 2020 13:53:06 (System) []        at dmg.cells.nucleus.CellNucleus.lambda$wrapLoggingContext$5(CellNucleus.java:764)
07 Feb 2020 13:53:06 (System) []        at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
07 Feb 2020 13:53:06 (System) []        at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
07 Feb 2020 13:53:06 (System) []        at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
07 Feb 2020 13:53:06 (System) []        at org.dcache.util.BoundedExecutor$Worker.run(BoundedExecutor.java:251)
07 Feb 2020 13:53:06 (System) []        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
07 Feb 2020 13:53:06 (System) []        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
07 Feb 2020 13:53:06 (System) []        at dmg.cells.nucleus.CellNucleus.lambda$wrapLoggingContext$4(CellNucleus.java:755)
07 Feb 2020 13:53:06 (System) []        at java.lang.Thread.run(Thread.java:745)
07 Feb 2020 13:53:06 (System) [] Caused by: java.lang.IllegalArgumentException: Could not resolve placeholder 'some-phony-property' in value "${some-phony-property}"
07 Feb 2020 13:53:06 (System) []        at org.springframework.util.PropertyPlaceholderHelper.parseStringValue(PropertyPlaceholderHelper.java:178)
```

The domain, however, does not exit or restart:
```
alarmsDomain running (for 3 minutes) 27275 root /var/log/dcache/alarmsDomain.log 
```

Wrapping each line of the CellShell execution with System.err printouts, this is what I found:
```
CELL SHELL EXECUTING create org.dcache.alarms.logback.LogServerCell ${alarms.cell.name}          "classpath:org/dcache/alarms/logback/alarms.xml          -profiles=${alarms.db.type}          -consume=${alarms.cell.consume}"
07 Feb 2020 18:11:21 (alarms) [] Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanDefinitionStoreException: Invalid bean definition with name 'fake' defined in class path resource [org/dcache/alarms/logback/alarms.xml]: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"; nested exception is java.lang.IllegalArgumentException: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"
CELL SHELL ANSWER dmg.util.CommandThrowableException: (3) Invalid bean definition with name 'fake' defined in class path resource [org/dcache/alarms/logback/alarms.xml]: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"; nested exception is java.lang.IllegalArgumentException: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"
07 Feb 2020 18:11:21 (System) [] URL [file:/usr/share/dcache/services/alarms.batch]: line 45: Command failed: (3) Invalid bean definition with name 'fake' defined in class path resource [org/dcache/alarms/logback/alarms.xml]: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"; nested exception is java.lang.IllegalArgumentException: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"; caused by: org.springframework.beans.factory.BeanDefinitionStoreException: Invalid bean definition with name 'fake' defined in class path resource [org/dcache/alarms/logback/alarms.xml]: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"; nested exception is java.lang.IllegalArgumentException: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"
...
CELL SHELL DID NOT PROPAGATE ERROR FOR dmg.util.CommandThrowableException: (3) Invalid bean definition with name 'fake' defined in class path resource [org/dcache/alarms/logback/alarms.xml]: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"; nested exception is java.lang.IllegalArgumentException: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"
```

That is, the command answer was a CommandThrowableException.  Because this is not a CommandExitException, it is not propagated, so the domain finishes the call to execute peaceably.   

This is the place in the CellShell create command in question:
```
           } catch (ExecutionException e) {
               Throwables.throwIfInstanceOf(e.getCause(), CommandException.class);
               throw new CommandThrowableException(e.getCause().getMessage(), e.getCause());
           }
```
However, it is insufficient to capture the cause of this ExecutionException and throw a CommandExitException instead of the CommandThrowableException.   What seems to be required is the following:

```
          } catch (ExecutionException e) {
               Throwable c = e.getCause();
               if (c instanceof RuntimeException) {
                      _doOnError = ErrorAction.SHUTDOWN;
               }
               Throwables.throwIfInstanceOf(c, CommandException.class);
               throw new CommandThrowableException(c.getMessage(), c);
           }
```
Doing this causes the domain to exit and restart (which seems to be the behavior which is expected by the referrer of this issue in RT):

```
2020-02-07 19:00:07 Launching /usr/bin/java -server -Xmx4096m -XX:MaxDirectMemorySize=4096m -Dsun.net.inetaddr.ttl=1800 -Dorg.globus.tcp.port.range=20000,25000 -Dorg.dcache.dcap.port=0 -Dorg.dcache.ftp.log-aborted-transfers=true -Dorg.dcache.net.tcp.portrange=33115:33145 -Djava.security.krb5.realm=FNAL.GOV -Djava.security.krb5.kdc= -Djavax.security.auth.useSubjectCredsOnly=false -Djava.security.auth.login.config=/etc/dcache/jgss.conf -Dcontent.types.user.table=/etc/dcache/content-types.properties -Dzookeeper.sasl.client=false -Dcurator-dont-log-connection-problems=true -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/var/log/dcache/alarmsDomain-oom.hprof -XX:+UseCompressedOops -javaagent:/usr/share/dcache/classes/aspectjweaver-1.9.2.jar -Djava.awt.headless=true -DwantLog4jSetup=n -Djava.security.properties=/etc/dcache/max.java.security -Djdk.tls.ephemeralDHKeySize=2048 -Ddcache.home=/usr/share/dcache -Ddcache.paths.defaults=/usr/share/dcache/defaults org.dcache.boot.BootLoader start alarmsDomain
07 Feb 2020 19:00:10 (System) [] Running in ZooKeeper 3.4.x compatibility mode
07 Feb 2020 19:00:10 (System) [] Using emulated InjectSessionExpiration
07 Feb 2020 19:00:11 (alarms) [] Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanDefinitionStoreException: Invalid bean definition with name 'fake' defined in class path resource [org/dcache/alarms/logback/alarms.xml]: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"; nested exception is java.lang.IllegalArgumentException: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"
07 Feb 2020 19:00:11 (System) [] Failure at startup: (666) URL [file:/usr/share/dcache/services/alarms.batch]: line 45: (3) Invalid bean definition with name 'fake' defined in class path resource [org/dcache/alarms/logback/alarms.xml]: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"; nested exception is java.lang.IllegalArgumentException: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"
2020-02-07 19:00:12 Process terminated unexpectedly
2020-02-07 19:00:12 Sleeping 10 seconds

2020-02-07 19:00:22 Restarting /usr/bin/java -server -Xmx4096m -XX:MaxDirectMemorySize=4096m -Dsun.net.inetaddr.ttl=1800 -Dorg.globus.tcp.port.range=20000,25000 -Dorg.dcache.dcap.port=0 -Dorg.dcache.ftp.log-aborted-transfers=true -Dorg.dcache.net.tcp.portrange=33115:33145 -Djava.security.krb5.realm=FNAL.GOV -Djava.security.krb5.kdc= -Djavax.security.auth.useSubjectCredsOnly=false -Djava.security.auth.login.config=/etc/dcache/jgss.conf -Dcontent.types.user.table=/etc/dcache/content-types.properties -Dzookeeper.sasl.client=false -Dcurator-dont-log-connection-problems=true -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/var/log/dcache/alarmsDomain-oom.hprof -XX:+UseCompressedOops -javaagent:/usr/share/dcache/classes/aspectjweaver-1.9.2.jar -Djava.awt.headless=true -DwantLog4jSetup=n -Djava.security.properties=/etc/dcache/max.java.security -Djdk.tls.ephemeralDHKeySize=2048 -Ddcache.home=/usr/share/dcache -Ddcache.paths.defaults=/usr/share/dcache/defaults org.dcache.boot.BootLoader start alarmsDomain
07 Feb 2020 19:00:24 (System) [] Running in ZooKeeper 3.4.x compatibility mode
07 Feb 2020 19:00:24 (System) [] Using emulated InjectSessionExpiration
07 Feb 2020 19:00:25 (alarms) [] Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanDefinitionStoreException: Invalid bean definition with name 'fake' defined in class path resource [org/dcache/alarms/logback/alarms.xml]: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"; nested exception is java.lang.IllegalArgumentException: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"
07 Feb 2020 19:00:25 (System) [] Failure at startup: (666) URL [file:/usr/share/dcache/services/alarms.batch]: line 45: (3) Invalid bean definition with name 'fake' defined in class path resource [org/dcache/alarms/logback/alarms.xml]: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"; nested exception is java.lang.IllegalArgumentException: Could not resolve placeholder 'some.phony.property' in value "${some.phony.property}"
2020-02-07 19:00:25 Process terminated unexpectedly
2020-02-07 19:00:25 Sleeping 10 seconds

etc.
```

I do not know whether this is the correct solution to this problem, however, and would appreciate some input.

**Comment by alrossi** [comment]
Note also that I have confirmed this problem exists as far back as version 4.2.

**Comment by alrossi** [comment]
Changed the behavior of the initialize routine in ProxyDelegationService to throw FileNotFound error, expecting a 'continue' behavior.   Instead:   
```
10 Feb 2020 09:31:24 (Xrootd1-fndcatemp1) [] Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'server' defined in class path resource [org/dcache/xrootd/door/xrootd.xml]: Cannot resolve reference to bean 'channelhandler-factories' while setting bean property 'channelHandlerFactories'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'channelhandler-factories': FactoryBean threw exception on object creation; nested exception is java.io.FileNotFoundException: ProxyDelegationStore, vomDir /etc/grid-security/vomsdir
10 Feb 2020 09:31:24 (System) [] URL [file:/usr/share/dcache/services/xrootd.batch]: line 51: Command failed: (3) Failed to create bean 'server' : ProxyDelegationStore, vomDir /etc/grid-security/vomsdir; caused by: ProxyDelegationStore, vomDir /etc/grid-security/vomsdir
```
Same zombie behavior.

Also,

Here's another observation (aside from the fact that 'continue' is not working for some reason the way it should).   IN the batch file, the model seems to be
```
onerror shutdown
```
before the property checks, and then
```
onerror continue
```
for the actual create call.  The shell behavior is not set (onError is not written to) anywhere inside the Java code programmatically.  If we need to shutdown on a Runtime error in the bean config, then we need to do this (override continue).  I don't see any other way.

**Comment by alrossi** [comment]
The problem was in the batch script, using a 

```
onerror continue 
```

instead of 

```
onerror shutdown
```
directives just before the cell create call.

This prevents restart of the domain on fatal errors.

**Comment by kofemann** [comment]
fixed in c09ac4d
